Fix fnptr not auto implement fn once

badumbatish opened the pull request:
im a bit rather new to this area...

this fixes #3354 

gcc/rust/ChangeLog:

	* typecheck/rust-tyty.cc (VariantDef::clone): reformat 
	* (VariantDef::monomorphized_clone): reformat 
	* (FnPtr::setup_fn_once_output): Fix fnptr not auto implement fn once
	* typecheck/rust-tyty.h: Likewise.

gcc/testsuite/ChangeLog:

	* rust/compile/format_args_basic_expansion.rs: add sized and fn_once
	* rust/compile/issue-2042.rs: add sized and fn_once
	* rust/compile/privacy6.rs: add sized and fn_once
	* rust/compile/rust_abi.rs: add sized and fn_once
	* rust/compile/torture/function_reference2.rs: add sized and fn_once
	* rust/compile/torture/function_reference3.rs: add sized and fn_once
	* rust/compile/torture/function_reference4.rs: add sized and fn_once
	* rust/compile/try-catch-unwind-new.rs: add sized and fn_once
	* rust/compile/try-catch-unwind-old.rs: add sized and fn_once

badumbatish committed:
Fix fnptr not auto implement fn once

gcc/rust/ChangeLog:
	* typecheck/rust-tyty.cc (FnPtr::setup_fn_once_output):
	Fix fnptr not auto implement fn once
	* typecheck/rust-tyty.h: Likewise.

gcc/testsuite/ChangeLog:

	* rust/compile/format_args_basic_expansion.rs: add sized and fn_once
	* rust/compile/issue-2042.rs: add sized and fn_once
	* rust/compile/privacy6.rs: add sized and fn_once
	* rust/compile/rust_abi.rs: add sized and fn_once
	* rust/compile/torture/function_reference2.rs: add sized and fn_once
	* rust/compile/torture/function_reference3.rs: add sized and fn_once
	* rust/compile/torture/function_reference4.rs: add sized and fn_once
	* rust/compile/try-catch-unwind-new.rs: add sized and fn_once
	* rust/compile/try-catch-unwind-old.rs: add sized and fn_once

P-E-P commented on gcc/testsuite/rust/compile/try-catch-unwind-new.rs:
```diff
@@ -1,6 +1,19 @@
+
+
```
Those blank line should probably be removed ?

badumbatish replied:
Yes i'll fix this soon

P-E-P commented on gcc/testsuite/rust/compile/try-catch-unwind-old.rs:
```diff
@@ -1,6 +1,18 @@
+
```
Likewise.

badumbatish commented:
i somehow missed privacy6.rs ... will force push soo for merge if Philip doesn't have any change requests

dkm commented:
I may be wrong, but I think you're missing the "Signed-off-by" footer (unless you've filed copyright assignment, in that case please ignore my comment) :)

CohenArthur commented on gcc/rust/typecheck/rust-tyty.cc:
```diff
@@ -2222,6 +2222,56 @@ FnPtr::clone () const
+      return;
+    }
+
+  DefId &trait_id = fn_once_lookup.value ();
+  DefId &trait_item_id = fn_once_output_lookup.value ();
```
this would also allow you to avoid these lines, as `get_lang_item` already returns a `DefId` instead of an `optional<Defid>`

badumbatish replied:
i'll probably fix the original code from Philip too

CohenArthur reviewed (commented):
I think the code looks good, and it's not easy so well done. I'll let Philip take a look before merging as that is more his area of expertise. Thanks for working on this @badumbatish!!

philberty reviewed (commented):
I think this is part of the fix but there are some work to be done for bounds checking. Which i think the original issue is about. I think we do need this though.

badumbatish commented on gcc/rust/typecheck/rust-tyty.cc:
```diff
@@ -2222,6 +2222,56 @@ FnPtr::clone () const
 }
 
+void
+FnPtr::setup_fn_once_output () const
```
yes, will do

powerboat9 replied:
Would adding this to `CallableTypeInterface` work?

badumbatish replied:
could work, ill refactor

badumbatish replied:
i need to rebuild clangformat 10 on my ubuntu first...

powerboat9 replied:
tbh I just use clang format 19 -- it mostly works, aside from the once-in-6-months edge case

badumbatish replied:
it's refactoring some subparts of a source file and it's starting to be really a pain in the butt for me. I'm not sure if i should maintain some sort of repository to store different architecture of clang format 10...

powerboat9 replied:
Have you used `git add -p` and `git restore -p`?

badumbatish replied:
hey im back again! sorry hahah i just grad from uni

>Have you used git add -p and git restore -p

really helpful i havent considered it, tysm!

powerboat9 replied:
An OS course I took last fall semester had a lab with old machines that we worked on, which if I recall right had git `2.17`. `git restore` didn't exist in that version :)

badumbatish committed:
Consolidate setup_fn_once_output into superclass

gcc/rust/ChangeLog:

	* typecheck/rust-tyty.cc (CallableTypeInterface::setup_fn_once_output):
	Consolidate setup_fn_once_output into superclass.
	(FnPtr::setup_fn_once_output): likewise.
	(ClosureType::setup_fn_once_output): likewise.
	* typecheck/rust-tyty.h: likewise.

gcc/testsuite/ChangeLog:

	* rust/compile/format_args_extra_comma.rs: add sized and fn_once
	* rust/compile/issue-3615.rs: likewise.
	* rust/compile/issue-3647.rs: likewise.
	* rust/compile/issue-3654.rs: likewise.
